### ISSUE TITLE
cfncluster jobs not starting

### ISSUE THREAD

**Issue by karlvirgil** [awslabs/cfncluster#277]
I have a new cfncluster instance that is using slurm as its scheduler started with the following config file:

```
[aws]
aws_region_name = us-east-1
aws_access_key_id = ###
aws_secret_access_key = ###

[cluster fire]
ebs_settings = 2017-11-30-openfoam
vpc_settings = poc_vpn
key_name = fire
base_os = centos7
placement_group = fire_cluster
master_instance_type = c4.2xlarge
compute_instance_type = c5.18xlarge
initial_queue_size = 0
max_queue_size = 50
scheduler = slurm

[vpc poc_vpn]
master_subnet_id = subnet-a6a304fe
vpc_id = vpc-b53f65d1
use_public_ips = false

[ebs 2017-11-30-openfoam]
ebs_snapshot_id = snap-0d124431109028c5d

[global]
update_check = true
sanity_check = true
cluster_template = fire 
```

I launched one 3 node mpi job, and the job was started and shows up in squeue.  However, when I launch another 3 node job, the second job doesn't start running while the first job is running. One new extra c5.18xlarge compute node is created, but then no more are created so the second job doesn't start running, but just sits in the queue.

I created this cfncluster with version 1.4.0

Here's the output from squeue and sinfo:

```
(centos@ip-10-15-20-115)-(02:13 PM Thu Dec 14)-(/shared/simulations/slopedCeilingsAR/free-33deg)
foamdev $)sq
PENDING 31 /shared/simulations/slopedCeilingsAR/free-33deg-rotatedPallets
RUNNING 25 /shared/simulations/slopedCeilingsAR/free-33deg

centos
          ....centos....
  JOBID PARTITION     NAME     USER    STATE       TIME TIME_LIMI  NODES PRIORITY QOS NODELIST(REASON)
     31   compute ebyycncr   centos  PENDING       0:00 UNLIMITED      3 4294901753 (null) (Resources)
     25   compute ebyycncr   centos  RUNNING    6:55:41 UNLIMITED      3 4294901758 (null) ip-10-15-20-[35,55,166]


=======PARTITIONS======
PARTITION AVAIL  TIMELIMIT  NODES  STATE NODELIST
compute*     up   infinite      3  alloc ip-10-15-20-[35,55,166]
compute*     up   infinite      1   idle ip-10-15-20-103
```
Any thoughts?

-Karl

**Comment by mohanasudhan** [comment]
Correct my understanding if i am wrong

From your description, Your second job needs 3 nodes and you add 1 (c5.18xlarge) node. So idly  there is no enough resource to satisfy second job, hence job is waiting in queue. Once you have 3 idle node, job will start running.

**Comment by bwbarrett** [comment]
@karlvirgil, based on the fact that CfnCluster added one node, scaling appears to be working.  Note that CfnCluster's scaling behavior is pretty simple and its defaults pretty basic.  By default, if there are queued jobs (but less than 200 queued jobs), CfnCluster will add 1 new node every 300 seconds until there are no more queued jobs.  If there are more than 200 queued jobs, CfnCluster will add 10 new nodes every 300 seconds.  So in your example, it will take 2 minutes before the first node starts (because the queue alarm is for 2 consecutive 1 minute data points) and then 10 minutes for the next two nodes to start (5 minutes between each scale up), plus a bit for the last instance to actually start.  So between submitting the job and the job starting to run, it can take 15 minutes with everything going right.

In testing your config file under my account (obviously changing the vpc-id and whatnot), the cluster did eventually run the second 3 node job, but it took about the predicted 15 minutes before the job started.  Let me know if your cluster never gets to that point.

Obviously, this is not what you want.  The current CfnCluster behavior is a bit simplistic, scaling based on jobs instead of desired nodes.  But getting a consistent behavior across all three supported schedulers is a bit of a pain, so that's what CfnCluster does today.  The good news is that if you know something about your usual job sizes, you can control many pieces of the scaling behavior, as documented in http://cfncluster.readthedocs.io/en/latest/configuration.html#scaling.

For example, if your average job size is 5 nodes, it would make sense to change `scaling_adjustment` from the default of 1 to 5.  Now, whenever a job is waiting in the queue, CfnCluster will start 5 new instances instead of 1, reducing your wait time from 15 minutes to about 5.  You can fiddle with the `scaling_evaluation_period` and `scaling_cooldown` to try to make thing faster, but you may accidentally overshoot your scaling if they're too small.  The input into the scaling algorithm is a cron job that runs on the master node every 60 seconds, so there's really no point in making `scaling_period` less than 60.

Of course, you can also change the desired instances in the cluster's autoscaling group (either by modifying the config file and then running `cfncluster update <cluster>` or by changing the autoscaling group configuration directly to make the cluster scale up faster.

**Comment by karlvirgil** [comment]
@bwbarrett, thanks for your detailed explanation.  Makes sense now.  I was able to get to the bottom of the problem.  I was being limited to a maximum of 5 c5.18xlarge compute nodes.  I need get this limit changed to a higher value before more new nodes would start.  So, I submitted a ticket to get this limit increased, and then new nodes were added as expected.  Seems like there could be better diagnostics from within a cfncluster to see when these AWS limits have been reached, so the user doesn't need to wade through the AWS console tabs to figure it out.

Thanks again for your help.  I think that I'll tweak my settings so it is not so slow to add new nodes.

-Cheers, Karl
